Remove bbcode_text from RichTextLabel

HaSa1002 opened the pull request:
This simplyfies the text setting and resolves some bugs as the internal data structure is the same.

*It's a draft, since the points below must be discussed first (at least 1 to 3)*

Things to discuss:
1. Make add_text, set_text, get_text private (hide them in their current form)?
2. Rename set_bbcode, get_bbcode and append_bbcode to set_text, get_text and add/append_text?
3. Rename bbcode_enabled?
4. How to parse old scene files (since we should put the `bbcode_text` into the `text`, when `bbcode_enabled` is true)?

fixes #36711
fixes #35553 and therefore supersedes #36037 if merged
fixes partly #18413

#35368 seems related acording to the title, but issue itself is unclear

We could alternativly make the RichTextLabel a base class for multiple parsers and reimplement a BBCodeLabel or go in the direction of https://github.com/godotengine/godot-proposals/issues/826

HaSa1002 committed:
Remove bbcode_text from RichTextLabel

Also renames:
 - append_bbcode -> append_text
 - get_bbcode -> get_text
 - set_bbcode -> set_text
 - get_text -> get_parsed_text

Property text is:
set_text
get_text

KoBeWi commented:
Current `get_text()` is actually useful, because it returns the actual text, without bbcodes. This possibility should be preserved somehow.

HaSa1002 commented:
I thought about having get_parsed_text, or rename the the bbcode text to something like get_raw_text

HaSa1002 commented:
Merged set_text
Renamed
 - append_bbcode -> append_text
 - get_bbcode -> get_raw_text
 - set_bbcode -> set_text

text property is now set_text and get_raw_text. Maybe a change to get_text and rename of get_text to get_parsed_text is desired for consistency.

HaSa1002 commented:
rebased and tested.
I don't like having made the text property to `set_text`and `get_raw_text`, but I don't know how to phrase `get_text_but_without_bbcode`. Any ideas?

KoBeWi commented on doc/classes/RichTextLabel.xml:
```diff
@@ -101,6 +101,13 @@
+			<return type="String">
+			</return>
+			<description>
+				Returns the raw text, without Bbcode parsing.
```
```suggestion
				Returns the raw text, without BBCode parsing.
```

KoBeWi commented:
The `_validate_property()` can be removed.

As for the naming, getter for text should be called `get_text()`. I feel like the `get_text()` and `get_raw_text()` would be fine swapped 🤔 Other names for no-BBCode could be `get_clean_text()`, `get_final_text()`, `get_unformatted_text()` or `get_parsed_text()`.

Also the internal variable `bbcode` could be renamed to `text`, because it introduces some confusion.

HaSa1002 commented:
> The `_validate_property()` can be removed.

done.

> As for the naming, getter for text should be called `get_text()`. I feel like the `get_text()` and `get_raw_text()` would be fine swapped 🤔 Other names for no-BBCode could be `get_clean_text()`, `get_final_text()`, `get_unformatted_text()` or `get_parsed_text()`.

I renamed `get_raw_text()` to `get_parsed_text()`. However, `get_text()` returns the raw text, if `bbcode_enabled` is true and the parsed text if not. Is that desired? It looks like a hack for the text <--> bbcode_text ambigity. I deleted that.

> Also the internal variable `bbcode` could be renamed to `text`, because it introduces some confusion.

done.

KoBeWi commented on doc/classes/RichTextLabel.xml:
```diff
@@ -80,6 +80,13 @@
+			<return type="String">
+			</return>
+			<description>
+				Returns the BBCode text without BBCode mark-up.
```
I'd change the description to `Returns the text without BBCode mark-up.` or `Returns the text with BBCode mark-up removed.`

KoBeWi commented on doc/classes/RichTextLabel.xml:
```diff
@@ -491,8 +494,8 @@
-			The raw text of the label.
-			When set, clears the tag stack and adds a raw text tag to the top of it. Does not parse BBCodes. Does not modify [member bbcode_text].
+			The label's text in BBCode format. Is not representative of manual modifications to the internal tag stack. Erases changes made by other methods when edited.
+			[b]Note:[/b] It is unadvised to use the [code]+=[/code] operator with [code]text[/code] (e.g. [code]text += "some string"[/code]) as it replaces the whole text and can cause slowdowns. Use [method append_text] for adding text instead, unless you absolutely need to close a tag that was opened in an earlier method call.
```
The note is only relevant when `bbcode_enabled` is true. It should probably mention it.

HaSa1002 replied:
done

KoBeWi reviewed (approved):
Looks alright now.

KoBeWi commented on doc/classes/RichTextLabel.xml:
```diff
@@ -491,8 +494,8 @@
-			The raw text of the label.
-			When set, clears the tag stack and adds a raw text tag to the top of it. Does not parse BBCodes. Does not modify [member bbcode_text].
+			The label's text in BBCode format. Is not representative of manual modifications to the internal tag stack. Erases changes made by other methods when edited.
+			[b]Note:[/b] If [member bbcode_enabled] is true it is unadvised to use the [code]+=[/code] operator with [code]text[/code] (e.g. [code]text += "some string"[/code]) as it replaces the whole text and can cause slowdowns. Use [method append_text] for adding text instead, unless you absolutely need to close a tag that was opened in an earlier method call.
```
```suggestion
			[b]Note:[/b] If [member bbcode_enabled] is [code]true[/code], it is unadvised to use the [code]+=[/code] operator with [code]text[/code] (e.g. [code]text += "some string"[/code]) as it replaces the whole text and can cause slowdowns. Use [method append_text] for adding text instead, unless you absolutely need to close a tag that was opened in an earlier method call.
```

reduz commented:
Is this the final resolution about this that everyone agrees to? I remember too many discussions on this topic at this point.

reduz commented:
Ok I got confirmation that this is the case, so feel free to rebase and merge.

HaSa1002 commented:
will do

HaSa1002 commented:
Rebased. One thing to remember: This **will** erase all text from RTLs as they are not saved in the `text` property. Can we add a property upgrade hint somehow?

akien-mga commented:
> Rebased. One thing to remember: This will erase all text from RTLs as they are not saved in the text property. Can we add a property upgrade hint somehow?

Yes, see e.g. this:
https://github.com/godotengine/godot/blob/104a6191fd32ca7bd28b28df6de091652c3d92cd/scene/resources/environment.cpp#L1032-L1049

HaSa1002 commented:
> 
> 
> > Rebased. One thing to remember: This will erase all text from RTLs as they are not saved in the text property. Can we add a property upgrade hint somehow?
> 
> Yes, see e.g. this:
> 
> https://github.com/godotengine/godot/blob/104a6191fd32ca7bd28b28df6de091652c3d92cd/scene/resources/environment.cpp#L1032-L1049

Thank you. Done and rebased

akien-mga commented:
Thanks!
